### PR TITLE
Make compatible with pymbar 3 and 4

### LINKED ISSUE #629: Compatibility with pymbar 4
Are there plans to update this to work with the new pymbar? Is this something that I could help out with?

### PR DESCRIPTION
**Pull request by mattwthompson**
## Description
This PR attempts to introduce compatibility with pymbar 4 in a way that does not break compatibility with pymbar 3. This is heavily inspired by https://github.com/shirtsgroup/physical_validation/pull/220

Resolves #629 

## Todos
- [ ] Implement feature / fix bug
- [ ] Add [tests](https://github.com/choderalab/openmmtools/tree/master/openmmtools/tests)
- [ ] Update [documentation](https://github.com/choderalab/openmmtools/tree/master/docs) as needed
- [ ] Update [changelog](https://github.com/choderalab/openmmtools/blob/master/docs/releasehistory.rst) to summarize changes in behavior, enhancements, and bugfixes implemented in this PR

## Status
- [ ] Ready to go

## Changelog message
```

```

### CODE REVIEW

**Comment by mikemhenry** [comment]
@mattwthompson Looking at this now, what is left to do?

**Comment by mikemhenry** [comment]
@mattwthompson I will give you write access to this repo, then you can push this branch directly so openeye will work. Or I can just take over the PR and push it myself, let me know what you rather do

**Comment by ijpulidos** [comment]
We shouldn't be needing openeye for openmmtools

File: openmmtools/multistate/multistateanalyzer.py [hunk 1]
```diff
@@ -1915,11 +1920,13 @@ def _compute_free_energy(self):
         logger.debug("Computing covariance matrix...")
 
         try:
-            # pymbar 2
-            (Deltaf_ij, dDeltaf_ij) = self.mbar.getFreeEnergyDifferences()
-        except ValueError:
             # pymbar 3
             (Deltaf_ij, dDeltaf_ij, _) = self.mbar.getFreeEnergyDifferences()
```
Comment (ijpulidos) [inline openmmtools/multistate/multistateanalyzer.py:1924]: This should fix the failing tests.
```suggestion
            Deltaf_ij, dDeltaf_ij = self.mbar.getFreeEnergyDifferences()
```
pymbar 3 actually returns two things by default, unless we specify `return_theta=True`.

**Review by ijpulidos — CHANGES_REQUESTED** [review]
Minor code change to fix the tests. Pymbar 3 `getFreeEnergyDifferences` should actually return 2 objects by default.

**Comment by mattwthompson** [comment]
Sorry, I let the invitation expire by not getting to it in time. These changes shouldn't touch OpenEye but I can re-open this to an upstream branch if it ends up being needed.

It looks like it's only needed for generating data for tests, not actual API calls:

```console
$ grep -r openeye .
./openmmtools/testsystems.py:    molecule : openeye.oechem.OEMol
./openmmtools/testsystems.py:    from openeye import oechem
Binary file ./openmmtools/__pycache__/testsystems.cpython-39.pyc matches
./openmmtools/data/benzene-toluene-implicit/generate-molecules.py:from openeye import oechem
./openmmtools/data/benzene-toluene-implicit/generate-molecules.py:from openeye import oeomega
./openmmtools/data/benzene-toluene-implicit/generate-molecules.py:from openeye import oeiupac
./openmmtools/data/benzene-toluene-implicit/generate-molecules.py:from openeye import oequacpac
./openmmtools/data/cb7-viologen/molecules/create-viologen.py:from openeye.oechem import *
./openmmtools/data/cb7-viologen/molecules/create-viologen.py:from openeye.oeomega import *
./openmmtools/data/cb7-viologen/molecules/create-viologen.py:from openeye.oequacpac import *

**Comment by ijpulidos** [comment]
Thanks a lot @mattwthompson , this is looking good! Please let us know if this is good to be merged

**Comment by mattwthompson** [comment]
I think this is good to go - in the sense that I implemented what I set out to do originally. The log of these tests is pretty much meaningless to me so I'm not sure if these tests are actually passing or not?

**Comment by ijpulidos** [comment]
@mattwthompson I think the fails in the tests are unrelated with the changes in this PR. More info at #654 . Once we review and merge that one, tests here should run fine. Thanks!

**Comment by ijpulidos** [comment]
We are hitting random fails in the tests as in https://github.com/choderalab/openmmtools/actions/runs/4227354498/jobs/7341747889#step:9:3130

I can reproduce this behavior locally, sometimes it passes sometimes it fails when using `pymbar=4.0.1`.  Whereas it seems to pass every time I run them with `pymbar=3.1.1`

**Comment by ijpulidos** [comment]
Oh even though we are testing with both versions and they both seem to be failing here in github actions. Locally they only fail with pymbar=4.0.1 for me.

**Comment by ijpulidos** [comment]
This was superseded byt #659 . Closing. Thanks @mattwthompson for all the work!

**Pull request closed without merging**